feat(enquiry): add recaptcha

nauynix opened the pull request:
Adds recaptcha to enquiry form. Server recaptcha code under `services` is adapted from `formSG`.
The two new env variables are test keys given by google in the [public](https://developers.google.com/recaptcha/docs/faq#id-like-to-hide-the-recaptcha-badge.-what-is-allowed) 

Closes #17 

## Before & After Screenshots

**BEFORE**:
<img width="594" alt="image" src="https://user-images.githubusercontent.com/20250559/129695831-bf228e9e-8402-47a7-b96e-c6df213159e8.png">

**AFTER**:
<img width="593" alt="image" src="https://user-images.githubusercontent.com/20250559/129695752-d3fe3d4a-8eaa-4fa3-8d60-e27d7342d807.png">

**New environment variables**:

- `REACT_APP_RECAPTCHA_SITE_KEY` : Recaptcha key for client to generate user response
- `RECAPTCHA_SECRET_KEY` : Recaptcha key for server to verify user response

**New dependencies**:

`Client`
- `react-google-recaptcha`

`Server`
- `axios` : to verify user response for recaptcha
- `https-status-codes` : since the code for `formSG` uses this package too, good to add it now and change all the other handwritten status code 
- `neverthrow` : since the code for `formSG` uses this package too, good to add it now and change all the other return values in `services` to be `result`\`resultAsync`
- 'supertest` : to test `enquiry.routes` and `enquiry.controller`.

nauynix committed:
chore: add react-google-recaptcha

nauynix committed:
feat(enquiry): add and send recaptcha

nauynix committed:
chore(server): install axios and neverthrow dep

nauynix committed:
chore(server): install http-status-codes dep

nauynix committed:
feat: add captcha service and testing

nauynix committed:
feat(enquiry): verify captcha in server

nauynix committed:
feat(enquiry): verify captcha in server

nauynix committed:
chore(enquiry): use recaptcha site key from env

nauynix committed:
chore(recaptcha): add comment to test

nauynix committed:
chore(recaptcha): load test env var

LoneRifle committed:
fix(recaptcha): correct README, remove unneeded external script

- Ensure the correct env var is specified in README
- Rely just on react-google-recaptcha to provide the recaptcha,
  dropping the need to use Google's official script

LoneRifle commented on server/src/services/recaptcha/recaptcha.service.ts:
```diff
@@ -0,0 +1,55 @@
+} from './recaptcha.errors'
+
+const logger = createLogger(module)
+export const verifyCaptchaResponse = (
```
For consistency, we should have a RecaptchaService which accepts an axios client in its ctor. This would allow you to avoid the module mocking in your tests. Since you would also be injecting the RecaptchaService into the EnquiryController, you can also provide unit-testing for the controller and even the route (especially given that you now have supertest).

For convenience, the ctor can also accept just the Google ReCAPTCHA URL; you then construct an axios client from that

nauynix replied:
Going with the first suggestion of passing an axios into RecaptchaService

LoneRifle commented on server/src/services/recaptcha/recaptcha.constants.ts:
```diff
@@ -0,0 +1,2 @@
+export const GOOGLE_RECAPTCHA_URL =
+  'https://www.google.com/recaptcha/api/siteverify'
```
should be provided as a convict config whose sole value (or default) is this

LoneRifle commented on server/src/services/recaptcha/__tests__/recaptcha.service.spec.ts:
```diff
@@ -0,0 +1,85 @@
+} from '../recaptcha.errors'
+import { verifyCaptchaResponse } from '../recaptcha.service'
+
+const MOCK_PRIVATE_KEY = recaptchaConfig.recaptchaSecretKey
```
you're using mocks, so there's no need to reference bootstrap here.

LoneRifle reviewed (changes requested):
The biz logic is sound, having tested on dev and staging. Some feedback given concerning refactoring and providng coverage tests. Good work nonetheless!

LoneRifle commented on server/src/bootstrap/config/recaptcha.ts:
```diff
@@ -0,0 +1,26 @@
+}
+
+const recaptchaSchema: Schema<RecaptchaConfig> = {
+  recaptchaSiteKey: {
+    doc: 'Recaptcha key for client to generate user response',
+    format: String,
+    default: '',
+    env: 'REACT_APP_RECAPTCHA_SITE_KEY',
+  },
```
this is not actually needed here, since it's webpacked into the frontend

nauynix committed:
refactor: shift google url to convict

LoneRifle committed:
build(ci): put reCAPTCHA site key in build, not server

- remove `recaptchaSiteKey` from server config, not used
- supply `REACT_APP_RECAPTCHA_SITE_KEY` as a docker build arg
- have GitHub Actions select the env-dependent key at build time

LoneRifle committed:
fix(csp): add Google to frame-src for recaptcha

nauynix committed:
refactor: remove unneccesary convict of recaptcha site key

nauynix committed:
refactor: dependency injection of captcha

LoneRifle committed:
fix(enquiries): provide human-readable value for sender

LoneRifle committed:
fix(enquiries): provide human-readable value for sender

nauynix committed:
update recaptcha test to mock axios

Signed-off-by: nauynix <nauynix@gmail.com>

nauynix committed:
fix(enquiries): update test for human-readable value

Signed-off-by: nauynix <nauynix@gmail.com>

nauynix committed:
chore: install supertest in server dep

nauynix committed:
feat(enquiry): add test to controller

nauynix committed:
feat(enquiry): add testing for routes

nauynix committed:
Merge remote-tracking branch 'origin/feat/recaptcha' into feat/recaptcha

nauynix committed:
fix(enquiry): remove service test dependency on mail

LoneRifle commented on server/src/modules/enquiry/__tests__/enquiry.routes.spec.ts:
```diff
@@ -0,0 +1,96 @@
+import { routeEnquiries } from '../enquiry.routes'
+import { EnquiryService } from '../enquiry.service'
+
+describe('/enquires', () => {
```
```suggestion
describe('/enquiries', () => {
```

LoneRifle reviewed (approved):
lgtm otherwise, merge when ready

nauynix committed:
refactor(enquiry route test): correct spelling mistake

Co-authored-by: LoneRifle <LoneRifle@users.noreply.github.com>
